File: app/src/main/java/com/mirogal/cocktail/data/repository/CocktailRepository.java
package com.mirogal.cocktail.data.repository;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import com.mirogal.cocktail.data.database.CocktailDao;
import com.mirogal.cocktail.data.database.CocktailDatabase;
import com.mirogal.cocktail.data.database.entity.CocktailDbEntity;
import com.mirogal.cocktail.data.repository.netpagedlist.BoundaryCallback;
import com.mirogal.cocktail.data.repository.netpagedlist.DataSourceFactory;

public class CocktailRepository {

    private LiveData<PagedList<CocktailDbEntity>> saveCocktailList;
    private LiveData<PagedList<CocktailDbEntity>> selectCocktailList;
    private MutableLiveData<NetworkState.Status> networkStatus;
    private MutableLiveData<String> requestQuery;

    private CocktailDao cocktailDao;
    private DataSourceFactory dataSourceFactory;
    private BoundaryCallback boundaryCallback;

    private static CocktailRepository INSTANCE_REPOSITORY;

    public static CocktailRepository getInstance(final Context context) {
        if (INSTANCE_REPOSITORY == null) {
            synchronized (CocktailDatabase.class) {
                if (INSTANCE_REPOSITORY == null) {
                    INSTANCE_REPOSITORY = new CocktailRepository(context);
                }
            }
        }
        return INSTANCE_REPOSITORY;
    }

    private CocktailRepository(Context context) {
        selectCocktailList = new MutableLiveData<>();
        CocktailDatabase db = CocktailDatabase.getInstance(context);
        cocktailDao = db.getDao();
        networkStatus = new MutableLiveData<>();
        requestQuery = new MutableLiveData<>();
        initSaveCocktailList();
        initSelectCocktailList();
    }

    private void initSaveCocktailList() {
        PagedList.Config pagedListConfig = new PagedList.Config.Builder()
                .setPageSize(10)
                .setPrefetchDistance(30)
                .setInitialLoadSizeHint(30)
                .setEnablePlaceholders(false)
                .build();

        saveCocktailList = new LivePagedListBuilder<>(cocktailDao.getCocktailList(), pagedListConfig)
                .build();
    }

    private void initSelectCocktailList() {
        PagedList.Config pagedListConfig = new PagedList.Config.Builder()
                .setPageSize(10)
                .setPrefetchDistance(30)
                .setInitialLoadSizeHint(30)
                .setEnablePlaceholders(false)
                .build();

        dataSourceFactory = new DataSourceFactory();
        boundaryCallback = new BoundaryCallback(this);

        selectCocktailList = Transformations.switchMap(requestQuery, input -> {

            if (requestQuery.getValue() == null || requestQuery.getValue().isEmpty()) {
                dataSourceFactory.setCurrentQuery(null);
            } else {
                dataSourceFactory.setCurrentQuery(requestQuery.getValue());
            }
            dataSourceFactory.create();

            return new LivePagedListBuilder<>(dataSourceFactory, pagedListConfig)
                    .setBoundaryCallback(boundaryCallback)
                    .build();
        });
    }

    public LiveData<PagedList<CocktailDbEntity>> getSaveCocktailList() {
        return saveCocktailList;
    }

    public LiveData<PagedList<CocktailDbEntity>> getSelectCocktailList() {
        return selectCocktailList;
    }

    public MutableLiveData<NetworkState.Status> getNetworkStatus() {
        return networkStatus;
    }

    public MutableLiveData<String> getRequestQuery() {
        return requestQuery;
    }

    public void saveCocktail(CocktailDbEntity cocktail) {
        new Thread(new Runnable() {
            public void run() {
                cocktailDao.insertCocktail(cocktail);
            }
        }).start();
    }

    public void deleteCocktail(int cocktailId) {
        new Thread(new Runnable() {
            public void run() {
                cocktailDao.deleteCocktail(cocktailId);
            }
        }).start();
    }
}

File: app/src/main/java/com/mirogal/cocktail/data/repository/CocktailRepository.kt
package com.mirogal.cocktail.data.repository

import android.app.Application
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.mirogal.cocktail.data.db.CocktailDatabase
import com.mirogal.cocktail.data.db.entity.CocktailDbEntity
import com.mirogal.cocktail.data.repository.netpagedlist.BoundaryCallback
import com.mirogal.cocktail.data.repository.netpagedlist.DataSourceFactory

class CocktailRepository(application: Application) {

    private val database = CocktailDatabase.newInstance(application)

    lateinit var saveCocktailListLiveData: LiveData<List<CocktailDbEntity>?>
    var loadCocktailListLiveData: LiveData<PagedList<CocktailDbEntity?>> = MutableLiveData()
    val searchNameMutableLiveData: MutableLiveData<String?> = MutableLiveData()
    val networkStatusMutableLiveData: MutableLiveData<NetworkState.Status> = MutableLiveData()

    companion object {

        @Volatile
        private var INSTANCE: CocktailRepository? = null

        fun newInstance(application: Application): CocktailRepository {
            val tempInstance = INSTANCE
            if (tempInstance != null) {
                return tempInstance
            }
            synchronized(this) {
                val instance = CocktailRepository(application)
                INSTANCE = instance
                return instance
            }
        }
    }

    init {
        initSaveCocktailList()
        initLoadCocktailList()
    }

    private fun initSaveCocktailList() {
        saveCocktailListLiveData = database.dao().cocktailListLiveData
    }

    private fun initLoadCocktailList() {
        val pagedListConfig = PagedList.Config.Builder()
                .setPageSize(20)
                .setPrefetchDistance(40)
                .setInitialLoadSizeHint(40)
                .setEnablePlaceholders(false)
                .build()
        val dataSourceFactory = DataSourceFactory()
        val boundaryCallback = BoundaryCallback(this)
        loadCocktailListLiveData = Transformations.switchMap(searchNameMutableLiveData) {
            if (searchNameMutableLiveData.value == null || searchNameMutableLiveData.value!!.isEmpty()) {
                dataSourceFactory.setCurrentQuery(null)
            } else {
                dataSourceFactory.setCurrentQuery(searchNameMutableLiveData.value)
            }
            dataSourceFactory.create()
            LivePagedListBuilder(dataSourceFactory, pagedListConfig)
                    .setBoundaryCallback(boundaryCallback)
                    .build()
        }
    }


    fun getCocktailById(cocktailId: Int): LiveData<CocktailDbEntity> {
        return database.dao().getCocktailById(cocktailId)
    }

    fun addCocktailToDb(cocktail: CocktailDbEntity?) {
        Thread(Runnable { database.dao().addCocktail(cocktail!!) }).start()
    }

    fun deleteCocktailFromDb(cocktailId: Int) {
        Thread(Runnable { database.dao().deleteCocktail(cocktailId) }).start()
    }

    fun setCocktailStateFavorite(cocktailId: Int, isFavorite: Boolean) {
        Thread(Runnable { database.dao().setFavorite(cocktailId, isFavorite) }).start()
    }

}